File: frontend/src/types/incident.ts
/**
 * TypeScript types for incidents matching backend API models
 */

// Incident type enum matching backend
export enum IncidentType {
  ARMED_ATTACK = 'armed_attack',
  INSURGENT_ATTACK = 'insurgent_attack',
  KIDNAPPING = 'kidnapping',
  BANDITRY = 'banditry',
  FARMER_HERDER_CLASH = 'farmer_herder_clash',
  CATTLE_RUSTLING = 'cattle_rustling',
  BOMB_BLAST = 'bomb_blast',
  ROBBERY = 'robbery',
  COMMUNAL_CLASH = 'communal_clash',
}

// Severity level enum
export enum SeverityLevel {
  LOW = 'low',
  MODERATE = 'moderate',
  HIGH = 'high',
  CRITICAL = 'critical',
}

// GeoJSON Point geometry
export interface PointGeometry {
  type: 'Point';
  coordinates: [number, number]; // [longitude, latitude]
}

// Casualties information
export interface Casualties {
  killed: number;
  injured: number;
  missing: number;
}

// Main Incident interface
export interface Incident {
  id: string;
  incident_type: IncidentType;
  severity: SeverityLevel;
  severity_score?: number;
  location: PointGeometry;
  location_name?: string;
  state?: string;
  lga?: string;
  description: string;
  timestamp: string; // ISO 8601 datetime string
  casualties?: Casualties;
  verified: boolean;
  verification_score: number;
  reporter_id?: string;
  reporter_phone?: string;
  is_anonymous?: boolean;
  media_urls?: string[];
  tags?: string[];
  created_at?: string;
  updated_at?: string;
  // Computed properties from backend
  latitude?: number;
  longitude?: number;
  distance_km?: number; // For nearby search results
}

// Incident creation payload
export interface IncidentCreate {
  incident_type: IncidentType;
  severity: SeverityLevel;
  location: PointGeometry;
  description: string;
  timestamp: string;
  casualties?: Casualties;
  reporter_phone?: string;
  is_anonymous?: boolean;
  media_urls?: string[];
  tags?: string[];
}

// Incident update payload
export interface IncidentUpdate {
  severity?: SeverityLevel;
  description?: string;
  verified?: boolean;
  casualties?: Casualties;
}

// Paginated list response
export interface IncidentListResponse {
  total: number;
  page: number;
  page_size: number;
  incidents: Incident[];
}

// Statistics response
export interface IncidentStats {
  total_incidents: number;
  by_type: Record<string, number>;
  by_severity: Record<string, number>;
  by_state: Record<string, number>;
  verified_count: number;
  unverified_count: number;
  casualties_total: {
    killed: number;
    injured: number;
    missing: number;
  };
  time_range_start: string;
  time_range_end: string;
  // Comparison metrics
  total_incidents_change?: number;
  verified_count_change?: number;
  casualties_change?: number;
  previous_period_total?: number;
}

// GeoJSON Feature for map display
export interface IncidentFeature {
  type: 'Feature';
  geometry: PointGeometry;
  properties: {
    id: string;
    incident_type: string;
    severity: string;
    severity_score: number;
    description: string;
    location_name: string;
    state: string;
    verified: boolean;
    verification_score: number;
    casualties: Casualties | null;
    timestamp: string;
  };
}

// GeoJSON FeatureCollection
export interface IncidentGeoJSON {
  type: 'FeatureCollection';
  features: IncidentFeature[];
}

// Query parameters for listing incidents
export interface IncidentListParams {
  page?: number;
  page_size?: number;
  incident_type?: IncidentType;
  severity?: SeverityLevel;
  state?: string;
  verified_only?: boolean;
  days?: number;
}

// Query parameters for nearby search
export interface NearbySearchParams {
  latitude: number;
  longitude: number;
  radius_km?: number;
  days?: number;
  incident_types?: string; // Comma-separated
  severities?: string; // Comma-separated
  verified_only?: boolean;
}

// Helper type for incident type labels
export const incidentTypeLabels: Record<IncidentType, string> = {
  [IncidentType.ARMED_ATTACK]: 'Armed Attack',
  [IncidentType.INSURGENT_ATTACK]: 'Insurgent Attack',
  [IncidentType.KIDNAPPING]: 'Kidnapping',
  [IncidentType.BANDITRY]: 'Banditry',
  [IncidentType.FARMER_HERDER_CLASH]: 'Farmer-Herder Clash',
  [IncidentType.CATTLE_RUSTLING]: 'Cattle Rustling',
  [IncidentType.BOMB_BLAST]: 'Bomb Blast',
  [IncidentType.ROBBERY]: 'Robbery',
  [IncidentType.COMMUNAL_CLASH]: 'Communal Clash',
};

// Helper type for severity labels
export const severityLabels: Record<SeverityLevel, string> = {
  [SeverityLevel.LOW]: 'Low',
  [SeverityLevel.MODERATE]: 'Moderate',
  [SeverityLevel.HIGH]: 'High',
  [SeverityLevel.CRITICAL]: 'Critical',
};

// Helper type for severity colors (Tailwind classes)
export const severityColors: Record<SeverityLevel, string> = {
  [SeverityLevel.LOW]: 'severity-low',
  [SeverityLevel.MODERATE]: 'severity-moderate',
  [SeverityLevel.HIGH]: 'severity-high',
  [SeverityLevel.CRITICAL]: 'severity-critical',
};

// Nigerian states list
export const NIGERIAN_STATES = [
  'Abia', 'Adamawa', 'Akwa Ibom', 'Anambra', 'Bauchi', 'Bayelsa',
  'Benue', 'Borno', 'Cross River', 'Delta', 'Ebonyi', 'Edo',
  'Ekiti', 'Enugu', 'Gombe', 'Imo', 'Jigawa', 'Kaduna',
  'Kano', 'Katsina', 'Kebbi', 'Kogi', 'Kwara', 'Lagos',
  'Nasarawa', 'Niger', 'Ogun', 'Ondo', 'Osun', 'Oyo',
  'Plateau', 'Rivers', 'Sokoto', 'Taraba', 'Yobe', 'Zamfara',
  'Federal Capital Territory',
] as const;

export type NigerianState = typeof NIGERIAN_STATES[number];

File: backend/app/schemas/incident.py
"""
Pydantic schemas for incident API validation and serialization
"""
from pydantic import BaseModel, Field, field_validator
from typing import Optional, List, Dict, Any
from datetime import datetime
from uuid import UUID
from app.models.incident import IncidentType, SeverityLevel


class PointGeometry(BaseModel):
    """GeoJSON Point geometry"""
    type: str = "Point"
    coordinates: List[float] = Field(..., min_length=2, max_length=2)

    @field_validator('coordinates')
    @classmethod
    def validate_coordinates(cls, v):
        """Validate longitude and latitude are in valid ranges"""
        if len(v) != 2:
            raise ValueError('Coordinates must be [longitude, latitude]')

        lon, lat = v
        if not (-180 <= lon <= 180):
            raise ValueError(f'Longitude must be between -180 and 180, got {lon}')
        if not (-90 <= lat <= 90):
            raise ValueError(f'Latitude must be between -90 and 90, got {lat}')

        return v


class CasualtyInfo(BaseModel):
    """Casualty information"""
    killed: int = Field(default=0, ge=0)
    injured: int = Field(default=0, ge=0)
    missing: int = Field(default=0, ge=0)


class IncidentCreate(BaseModel):
    """Schema for creating a new incident report"""
    incident_type: IncidentType
    severity: SeverityLevel
    location: PointGeometry
    description: str = Field(..., min_length=10, max_length=2000)
    timestamp: datetime
    casualties: Optional[CasualtyInfo] = None
    reporter_phone: Optional[str] = Field(None, max_length=20)
    is_anonymous: bool = False
    media_urls: Optional[List[str]] = None
    tags: Optional[List[str]] = None

    @field_validator('description')
    @classmethod
    def validate_description(cls, v):
        """Ensure description is meaningful"""
        if len(v.strip()) < 10:
            raise ValueError('Description must be at least 10 characters')
        return v.strip()


class IncidentUpdate(BaseModel):
    """Schema for updating an existing incident"""
    incident_type: Optional[IncidentType] = None
    severity: Optional[SeverityLevel] = None
    description: Optional[str] = None
    verified: Optional[bool] = None
    verification_notes: Optional[str] = None
    casualties: Optional[CasualtyInfo] = None
    media_urls: Optional[List[str]] = None
    tags: Optional[List[str]] = None


class IncidentResponse(BaseModel):
    """Schema for incident response"""
    id: UUID
    incident_type: IncidentType
    severity: SeverityLevel
    description: str
    location_name: Optional[str] = None
    state: Optional[str] = None
    lga: Optional[str] = None
    verified: bool
    verification_score: float
    casualties: Optional[Dict[str, int]] = None
    timestamp: datetime
    created_at: datetime
    updated_at: Optional[datetime] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    media_urls: Optional[List[str]] = None
    tags: Optional[List[str]] = None

    class Config:
        from_attributes = True


class IncidentGeoJSON(BaseModel):
    """GeoJSON feature for incident"""
    type: str = "Feature"
    geometry: PointGeometry
    properties: Dict[str, Any]


class IncidentListResponse(BaseModel):
    """Paginated list of incidents"""
    total: int
    page: int
    page_size: int
    incidents: List[IncidentResponse]


class NearbyIncidentsQuery(BaseModel):
    """Query parameters for nearby incidents search"""
    latitude: float = Field(..., ge=-90, le=90)
    longitude: float = Field(..., ge=-180, le=180)
    radius_km: float = Field(default=50.0, gt=0, le=500)
    days: int = Field(default=7, gt=0, le=365)
    incident_types: Optional[List[IncidentType]] = None
    severities: Optional[List[SeverityLevel]] = None
    verified_only: bool = False


class IncidentStats(BaseModel):
    """Statistics about incidents"""
    total_incidents: int
    by_type: Dict[str, int]
    by_severity: Dict[str, int]
    by_state: Dict[str, int]
    verified_count: int
    unverified_count: int
    casualties_total: Dict[str, int]
    time_range_start: datetime
    time_range_end: datetime
    # Comparison metrics
    total_incidents_change: Optional[float] = None  # Percentage change from previous period
    verified_count_change: Optional[float] = None
    casualties_change: Optional[float] = None
    previous_period_total: Optional[int] = None
